rskj-core/src/main/java/org/ethereum/crypto/signature/SignatureService.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * This file is part of RskJ
+ * Copyright (C) 2020 RSK Labs Ltd.
+ * (derived from ethereumJ library, Copyright (c) 2016 <ether.camp>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.ethereum.crypto.signature;
+
+import co.rsk.config.RskSystemProperties;
+import org.ethereum.crypto.ECKey;
+
+import javax.annotation.Nullable;
+import java.security.SignatureException;
+
+/**
+ * Class in charge of the implementations of all the Signature related functionality.
+ * Functions:
+ * - Sign data (future)
+ * - Recover PK from signature
+ * - Verify
+ *
+ * Is implemented as a Singleton, so the only way to access an instance is through getInstance().
+ */
+public abstract class SignatureService {
+
+    private static final String NATIVE_LIB = "native";
+
+    private static SignatureService INSTANCE;
+
+    /**
+     * <p> It should be called only once in Node Startup.</p>
+     *
+     * <p> It reads "crypto.library" property to decide which impl to initialize.</p>
+     *
+     * <p> By default it initialize Bouncy Castle impl.</p>
+     *
+     * @param rskSystemProperties
+     */
+    public static void initialize(RskSystemProperties rskSystemProperties) {
+        if (NATIVE_LIB.equals(rskSystemProperties.cryptoLibrary())) {//TODO: Check if Native library is loaded.
+            INSTANCE = new SignatureServiceNative();
+        } else {
+            INSTANCE = new SignatureServiceBC();
+        }
+    }
+    /**
+     * As a singleton this should be the only entry point for creating instances of SignatureService classes.
+     *
+     * @return either {@link SignatureServiceBC} or Native Signature (future) implementation.
+     */
+    public synchronized static final SignatureService getInstance() {
+        if (INSTANCE == null) {
```

Comment:
just an idea. Alternatively instead of creating the default implementation we may throw an exception if the service is not initialised (when INSTANCE == null). In this case I guess this method may not be `synchronized` which would have a little positive impact on the performance. Also I'd make `initialize` method `synchronized`.

Reply:
I change the implementation again. Tell me what you think of it.